### PR TITLE
Leaves - Elizabeth + Cloudy

### PR DESCRIPTION
**Pull request by north108**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | def + end, method signature (should there be parameters), code inside the block that does the actual thing you want.
What are the advantages of using git when collaboratively working on one code base? |Our work didn't necessitate that because we worked on one machine but we could use git to got back to a pervious changes. 
What kind of relationship did you and your pair have with the unit tests? |They helped us figure out what we needed to fix to get the code to work!
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | We used .map to put the words into their own string of characters and we used .min to find the word with the least amount of characters.
What was one method you and your pair used to debug code? | binding.pry, puts an p's 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We talked about how we would like to receive feedback, we also decided at the beginning if the project what the work flow was going to be like .

### CODE REVIEW

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
+require 'pry'
+
+
+def draw_letters
+  letters_in_hand = ARRAY_OF_LETTERS.shuffle.slice(1..10)
+  return letters_in_hand
+end
+
+ARRAY_OF_LETTERS = ["A", "A", "A", "A", "A", "A", "A", "A", "A", "B", "B", "C", "C", "D", "D", "D", "D", "E", "E", "E", "E", "E", "E", "E", "E", "E", "E", "E", "E", "F", "F", "G", "G", "G", "H", "H", "I", "I", "I", "I", "I", "I", "I", "I", "I", "J", "K", "L", "L", "L", "L", "M", "M", "N", "N", "N", "N", "N", "N", "O", "O", "O", "O", "O", "O", "O", "O", "P", "P", "Q", "R", "R", "R", "R", "R", "R", "S", "S", "S", "S", "T", "T", "T", "T", "T", "T", "U", "U", "U", "U", "V", "V", "W", "W", "X", "Y", "Y", "Z"]
```
Comment (beccaelenzil) [inline lib/adagrams.rb:9]: It's advisable to store the frequency information in a hash, and then use that hash to create an array. Also, it's best to put this global at the top of your script.


```ruby
letter_counts = {
  "A"=>9, "B"=>2, ...
}
letters = []
letter_counts.each do |letter, count|
  count.times do
    letters << letter
  end
end
```

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  
+  input.each do |chars|
+    if dup_letters_in_hand.include?(chars)
+      dup_letters_in_hand.delete(chars)
+    else 
+      return false
+    end
+  end
+  
+  check = 0
```
Comment (beccaelenzil) [inline lib/adagrams.rb:24]: This check seems redundant. The first loop will return false if a letter in the word is not in the hand. If that return doesn't happen, you can simply return true after that loop is finished.

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  end
+  return check == 0
+end
+
+def score_word(word)
+  letter_points = {
+    "A" => 1, "B" => 3, "C" => 3, "D" => 2, "E" => 1, "F" => 4, "G" => 2, "H" => 4, "I" => 1, "J" => 8, "K" => 5, "L" => 1, "M" => 3, "N" => 1, "O" => 1, "P" => 3, "Q"=> 10, "R"=> 1, "S" => 1, "T" => 1, "U" => 1, "V" => 4, "W" => 4, "X" => 8, "Y" => 4, "Z" => 10
+  }
+  word_array = word.upcase.chars
+  word_points = word_array.map do |k, v|
```
Comment (beccaelenzil) [inline lib/adagrams.rb:40]: Use meaningful names for k and v (i.e. letter and letter_value).

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    letter_points[k]
+  end
+  total_points = word_points.sum
```
Comment (beccaelenzil) [inline lib/adagrams.rb:43]: Consider using the .sum enumerable to combine the map and sum operations your perform.

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        winning_word = tiebreaker_length(winning_word, object)
+      end 
+    end
+    highest_score_hash[:word] = winning_word
+    highest_score_hash[:score] = score
+  end
+  return highest_score_hash
+end 
+
+def tiebreaker_length (old_word, new_word)
```
Comment (beccaelenzil) [inline lib/adagrams.rb:77]: Great use of a helper method. You could consider putting more of the tiebreaker functionality into this method.

**Comment by beccaelenzil** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions |yes
Small commits with meaningful commit messages |yes -- considering adding wave # to your commit messages
**Code Requirements** |
`draw_letters` method |
Uses appropriate data structure to store the letter distribution |see code review
All tests for `draw_letters` pass |yes
`uses_available_letters?` method |the check seems redundant
All tests for `uses_available_letters?` pass |yes
`score_word` method |
Uses appropriate data structure to store the letter scores |yes
All tests for `score_word` pass |yes
`highest_score_from` method |
Appropriately handles edge cases for tie-breaking logic |yes
All tests for `highest_score_from` pass |yes
**Overall** |Hey y'all! Great work on this project. Your code fulfills all of the requirements and does it in a readable and reasonable way. The code is clean, and it passes all the tests!



I'm making a few comments on this assignment about some suggestions on how to refactor. In particular, I call attention to some redundant code, suggest a way you can store letter frequency information in a data structure, and name your variables meaningfully to increase readability. Overall, you two have done a great job. Good work!
